src/azul/openapi.py

```diff
@@ -37,32 +109,52 @@ def annotated_specs(gateway_id, app, toplevel_spec) -> JSON:
     specs = aws.api_gateway_export(gateway_id)
     clean_specs(specs)
     specs = merge_dicts(toplevel_spec, specs, override=True)
-    docs = get_doc_specs(app)
-    for path, doc_spec in docs.items():
-        for verb, description in specs['paths'][path].items():
-            description = merge_dicts(description, doc_spec)
-            specs['paths'][path][verb] = description
-    return specs
+    path_specs, method_specs = get_app_specs(app)
+    return join_specs(specs, path_specs, method_specs)
 
 
 def clean_specs(specs):
     """
     Adjust specs from API Gateway so they pass linting
     """
     # Filter out 'options' since it causes linting errors
-    for path in specs['paths']:
-        specs['paths'][path].pop('options', None)
+    for path in specs['paths'].values():
+        path.pop('options', None)
     # Remove listed servers since API Gateway give false results
     specs.pop('servers')
 
 
-def get_doc_specs(app):
-    docs = {}
-    for route, entries in app.routes.items():
+def get_app_specs(app):
+    path_specs = {}
+    method_specs = {}
+    for entries in app.routes.values():
         func = next(iter(entries.values())).view_function
-        if hasattr(func, 'api_spec'):
-            docs[route] = func.api_spec
-    return docs
+        if hasattr(func, 'path_specs'):
+            for path, spec in func.path_specs.items():
+                if path in path_specs:
+                    assert spec == path_specs[path], f'Path spec for {spec} already exists'
+            path_specs.update(func.path_specs)
+        else:
+            # TODO (jesse): warn if route is undocumented
+            pass
+        if hasattr(func, 'method_specs'):
+            method_specs.update(func.method_specs)
+
+    return path_specs, method_specs
+
+
+def join_specs(toplevel_spec, path_specs, method_specs):
+    specs = toplevel_spec
+
+    for path, path_spec in path_specs.items():
+        # This may override entries from API Gateway with the same path
+        specs['paths'][path] = path_spec
+
+    for (path, method), method_spec in method_specs.items():
+        # This may override duplicate specs from path_specs or API Gateway
+        specs['paths'][path][method] = method_spec
+
+    return specs
```

Comment:
Looks like this code isn't covered by any tests: https://codecov.io/gh/DataBiosphere/azul/pull/1449/diff#D3-127.

Reply:
It is covered by the integration test. If you want better coverage, I can add more tests.

Reply:
*Some* of it is covered by IT. The error case isn't. Correct me if I am wrong but during my penultimate review the two-handlers-same-path case was not tested for and NOT handled. In response to that review you added a check for that to the code under test but the code is not present. So yes, I would like *unit* test coverage for at least the error case but ideally all of `get_app_specs`. 

Re `join_specs` not being covered by IT: it looks like it would be easy to do with a few doctests.

For future reference: when I say "coverage" or "covered" without qualification, I'm referring to the unit test coverage as measured by Coveralls and Codecov.

Reply:
Sorry for the multiple edits. Done now.

Reply:
Ok. I agree and I'll add this.